px.py

```diff
@@ -715,41 +744,111 @@ def do_CONNECT(self):
-            wlist = []
-            count = 0
-            max_idle = State.config.getint("settings", "idle")
-            while not State.exit:
-                count += 1
-                (ins, _, exs) = select.select(rlist, wlist, rlist, 1)
-                if exs:
-                    break
-                if ins:
-                    for i in ins:
-                        if i is self.proxy_socket:
-                            out = self.connection
-                        else:
-                            out = self.proxy_socket
-
-                        data = i.recv(4096)
-                        if data:
-                            out.send(data)
-                            count = 0
-                            cl += len(data)
-                if count == max_idle:
-                    break
-
-        dprint("Transferred %d bytes" % cl)
+            # Proxy connection may be already closed due to header (Proxy-)Connection: close
+            # received from proxy -> forward this to the client
+            if self.proxy_socket is None:
+                dprint("Proxy connection closed")
+                self.send_response(200, "True")
+                self.send_header("Proxy-Connection", "close")
+                self.end_headers()
+            else:
+                dprint("Tunneling through proxy")
+                self.send_response(200, "Connection established")
+                self.send_header("Proxy-Agent", self.version_string())
+                self.end_headers()
+
+                # sockets will be removed from these lists, when they are detected as closed by remote host;
+                # wlist contains sockets only when data has to be written
+                rlist = [self.connection, self.proxy_socket]
+                wlist = []
+
+                # data to be written to client connection and proxy socket respectively
+                cdata = []
+                sdata = []
+                idle = State.config.getint("settings", "idle")
+                max_idle = time.time() + idle
+                while not State.exit and (rlist or wlist):
+                    (ins, outs, exs) = select.select(rlist, wlist, rlist, idle)
+                    if exs:
+                        break
+                    if ins:
+                        for i in ins:
+                            if i is self.proxy_socket:
+                                out = self.connection
+                                wdata = cdata
+                                source = "proxy"
+                            else:
+                                out = self.proxy_socket
+                                wdata = sdata
+                                source = "client"
+
+                            data = i.recv(4096)
+                            if data:
+                                dlen = len(data)
+                                cl += dlen
+                                # Prepare data to send it later in the outs section
```

Comment:
`cl += len(data)`?

Reply:
Yes. Forgot to remove it after not using it anymore. :-)